### ISSUE TITLE
weakref и None

### ISSUE THREAD

**Issue by Lex-2008** [uliss4ever/My_repository_2#8]
Проблема в сеттере `prev_node` заключается в том, что если в него передать `None`, то он попытается сделать `self._prev_node = weakref.ref(None)`, на что weakref выкинет ошибку: "TypeError: cannot create weak reference to 'NoneType' object". То есть, None не поддерживается weakref'ом. Как же тут быть?

**Comment by uliss4ever** [comment]
Мне кажется, что вот так, обычную ссылку, но так же нужно в таком случае что-то сделать с геттером? Там либо return self._prev_node() либо return self._prev_node надо. Но как? Там же нет уже value.
Плюс ещё опять выпало, что такое value. Это очередная нода? следующий ребёнок? И вот эту проверку
 if value is not None and not isinstance(value, Node):
            raise ValueError
Когда предлагаемое value может оказаться не Нодой?
Специально пошла послушала Енота в этом месте, но он особо не поясняет.

![Снимок экрана 2021-03-08 в 18 51 25](https://user-images.githubusercontent.com/32711430/110348229-78ee1d80-8042-11eb-9daa-ad2a2aa229f5.png)

**Comment by Lex-2008** [comment]
Да, всё правильно!

В сеттере можно было бы условия переставить, чтобы было так:
* если value - None, то...
* иначе-если value - Node, то...
* иначе... (когда value - ни None, Node)

Может быть, будет покрасивше и попонятней. Но это не так важно.

По поводу геттера - посмотри внимательней на сеттер. Чем может быть `self._prev_node`? Тут есть два варианта (т.к. в сеттере у тебя в двух местах есть присваивание). В зависимости от того, что было в `value`, в `self._prev_node` попадают разные вещи. Какие?

**Comment by uliss4ever** [comment]
И даже вот это вынесу в отдельный вопрос (на завтра).
 Вот это вот (если оно правильно), что на детях или на грушах с яблоками означает?
 Потомсу что выше я не то чтоб без понимания совсем это сделала, просто я убрала слабую ссылку, если значение - None
       elif value is None:
            self._prev_node = value

**Comment by uliss4ever** [comment]
> Чем может быть self._prev_node? Тут есть два варианта (т.к. в сеттере у тебя в двух местах есть присваивание). В зависимости от того, что было в value, в self._prev_node попадают разные вещи. Какие?

Либо значение (value), либо значение по слабой ссылке . Но дело в том, я выше написала, я не до конца понимаю, что физически это значит. Вот есть список нод a-b-c-d Тогда value - это мы хотим приделать новую ноду e ? но проверка isinstance что проверяет? Как value  (то есть наша потенциальная нода е, может быть ни None, ни value?)

**Comment by Lex-2008** [comment]
> что такое value

Это то, что какие-то другие гипотетические программисты, пользующиеся нашим классом `Node`, пытаются присвоить в `prev_node`. Либо это защита от ошибок, чтобы мы в классе DoubleLinkedList не присвоили бы туда случайно то, чего не надо.

> ````
> if value is not None and not isinstance(value, Node):
>     raise ValueError
> ````
> Когда предлагаемое value может оказаться не Нодой?

Это проверка на ошибки. Чтобы если кто-то (мы или другие программисты) случайно попытается присвоить в prev_node не то, что надо, то это сразу же было видно. Иначе это может вызвать неожиданные ошибки когда-то потом, где-то в другом месте, когда, например, в `_get_node_` мы попытаемся перейти по ссылке prev_node, а там - уупс! вместо ноды окажется что-то другое. "_И кто ето туда положил?_" - воскликнем мы... Но будет уже поздно.

---

>> В зависимости от того, что было в value, в self._prev_node попадают разные вещи. Какие?
>
> Либо значение (value), либо значение по слабой ссылке

Правильно. В каком случае туда попадает value? На твоем скриншоте, что написано перед строчкой ` self._prev_node = value`?

**Comment by uliss4ever** [comment]
>В каком случае туда попадает value?

Если value == None получается.
А правильно я понимаю, что _prev_node = None если указатель на ноде а ? (а - это голова). Вот, я переставила вот так, правда isinstance пришлось убрать, а то пичарм ругался

![Снимок экрана 2021-03-08 в 19 55 51](https://user-images.githubusercontent.com/32711430/110353678-637bf200-8048-11eb-92de-af65910c365e.png)

**Comment by Lex-2008** [comment]
>> В каком случае туда попадает value?
> 
> Если value == None получается.

Правильно! Значит, если value == None, и мы делаем `self._prev_node = value`, то `self._prev_node` становится == None. Итак, ещё раз: какие вещи могут быть в `self._prev_node`?

Либо None, либо значение по слабой ссылке

И как мы в геттере можем отличить одно от другого? У нас же есть хорошая проверка, что можно написать после `if`, чтобы проверить, является ли нечто None или нет!

----

По поводу скриншота - все правильно, только если после elif вместо "value == Node" написать "isinstance(value, Node)" - пичарм ругаться не будет?

"value == Node" - это как спрашивать "_равна ли_ Буренка корове?", а "isinstance(value, Node)" - это "_является ли_ Буренка коровой?".

Разница в том, что и Буренка, и Машка _являются_ коровами, но при этом если Буренка _равна_ чему-то, и Машка _равна_ этому же самому, то получается, что Буренка и Машка _равны_ между собой, что совсем не обязательно!

**Comment by uliss4ever** [comment]
Даа, за них тревожно очень! Постоянно балансирую на грани, вспоминая горе-вошебника.
Не знаю, я тут не полную чушь написала? 

![Снимок экрана 2021-03-08 в 20 21 10](https://user-images.githubusercontent.com/32711430/110357141-04b87780-804c-11eb-8bde-291e9c7a548d.png)

**Comment by Lex-2008** [comment]
Сеттер мне нравится, геттер ещё не очень, но ты на правильном пути!

Что ты проверяешь, не None ли оно?

Куда мы записываем value, None, или слабую ссылку в сеттере?

Значит, откуда должен их брать геттер?

Ну и, если оно оказалось None, то значит, в сеттер было передано что? В каком случае сеттер сохраняет None?

**Comment by uliss4ever** [comment]
>В каком случае сеттер сохраняет None?

если value было None
Может тогда:
if self._prev_node is not None ? Потому что None или значение по слабой ссылке - у нас в self._prev_node записывается

![Снимок экрана 2021-03-08 в 20 42 36](https://user-images.githubusercontent.com/32711430/110359703-da1bee00-804e-11eb-95c4-5958240efbd8.png)

**Comment by Lex-2008** [comment]
Да, вот так совершенно правильно!
Я бы только вместо последнего `return` написал бы `return None`, чтобы было видно, что мы целенаправленно возвращаем значение None, а не просто завершаем функцию - но, похоже, в Питоне это совершенно идентичные конструкции.

Итак, подытожим: значит, сеттер для prev_node у нас сохраняет в `_prev_node` либо None, либо слабую ссылку на Node;
а геттер смотрит, что туда сеттер сохранил и, если там был None - то возвращает None, а иначе - Node, на которую ссылается слабая ссылка.
И всё!

Тебе более или менее понятно? Я не слишком сильно подсказывал?

**Comment by uliss4ever** [comment]
Ой, круто.. И я даже понимаю, спасибо тебе. Точнее так: Спасибо.
Да, в Питоне всё равно, но  пожалуй как раз тут и напишу, чтобы было ясно, что эта None-субстранция - это то, чего мы хотим. 
Я только единственно немного беспокоюсь, правильно ли думаю. _prev_node = None для "головы". То есть это мы в инсёрте, например, можем использовать?  
Или вообще напрямую? Просто задать ноду...

**Comment by uliss4ever** [comment]
>Тебе более или менее понятно? Я не слишком сильно подсказывал?
 
Боюсь сглазить (шутка), но такое чувство, что мне скорее более понятно, чем менее)) И да, несильно, мне кажется, потому что хотя у меня реально просвет в голове наступил, но пока при слове "сеттер" я немного пугаюсь. Но скорее по старой памяти, когда это ужас-ужас был...

**Comment by Lex-2008** [comment]
https://www.google.com/search?q=%D1%81%D0%B5%D1%82%D1%82%D0%B5%D1%80&tbm=isch А чего их бояться-то? :dog: 

Да, для "головы"  _prev_node = None. Но как это в инсёрте использовать - не очень понятно, т.к. "голова" в инсёрте доступна просто по ссылке `self.head`.

**Comment by uliss4ever** [comment]
Я вот ещё с листом_с_драйвером разбираюсь, как считаешь, достаточно ли корректно я попросила фабрику вернуть драйвер?

![Снимок экрана 2021-03-08 в 21 02 24](https://user-images.githubusercontent.com/32711430/110362020-d047ba00-8051-11eb-88c5-fc41ea950c14.png)

**Comment by uliss4ever** [comment]
>Но как это в инсёрте использовать - не очень понятно, т.к. "голова" в инсёрте доступна просто по ссылке self.head.

Да, правда. Просто я пытаюсь смоделировать для себя ситуацию, когда я подаю в сеттер None.
Они невозможно милые)))

**Comment by Lex-2008** [comment]
Нет, недостаточно корректно. Во-первых, посмотри самую предпоследнюю строчку в файле https://github.com/uliss4ever/Linkedlist/blob/branch/Fabric_method.py - как мы там метод `get_driver` вызываем? Особенно на ту часть, что в скобках.

Во-вторых, если написать:
```
a = ll.driver
b = ll.driver
c = ll.driver
```

то a, b, c - будут один и тот же драйвер или разные? А как должно быть? (подсказка: будет неправильно).

> Они невозможно милые)))

Ну так вот я и говорю!

**Comment by uliss4ever** [comment]
Вот эту строчку? driver = DriverFabric.get_driver() Да, я селф запихнула туда, не надо было. 
>a = ll.driver
b = ll.driver
c = ll.driver

По идее это один и тот же драйвер... А как должно быть, я пока не очень понимаю. Понимаю только что у нас он может быть один только какой-то, дефолтный, который мы выбрали (?) и с которым работаем. Мы, если можно, повызываем завтра это всё, в разных комбинациях? И в листе тоже. А то эту часть (использование того, что написано) я делаю не очень уверенно и не на 100%. Где-то вообще на логику надеюсь просто, а как применить эту штуку - не знаю.

**Comment by uliss4ever** [comment]
Да, у меня еще как-то странно в нашем с тобой проекте вот это место выглядит вот так (подчеркнутое), а в том, который я Еноту шлю - нормально (хотя файлы идентичны, сверила уже вчера построчно)
![Снимок экрана 2021-03-09 в 15 11 21](https://user-images.githubusercontent.com/32711430/110469139-531c5380-80ea-11eb-85c9-a9c5fcfedaff.png)
![Снимок экрана 2021-03-09 в 15 11 43](https://user-images.githubusercontent.com/32711430/110469143-53b4ea00-80ea-11eb-9357-a9f36b28cc3b.png)

 импортами у меня есть вопросы.

**Comment by Lex-2008** [comment]
Не знаю, должно подчеркивать. И предлагать исправление, если щелкнуть по красной лампочке:
![Screenshot_2021-03-09_15-21-41](https://user-images.githubusercontent.com/6545284/110485616-21a48780-80ec-11eb-93a3-87f8d7b330a0.png)

**Comment by uliss4ever** [comment]
Вот это пока не окончательно, но пусть тут будет)
![Снимок экрана 2021-03-09 в 20 12 48](https://user-images.githubusercontent.com/32711430/110510092-e918a400-8113-11eb-9efe-778a6c5b4d0e.png)

**Comment by Lex-2008** [comment]
Да, хорошо, только у тебя там бесконечная рекурсия получилась (когда функция вызывает сама себя). `self.driver` (который без `_`) - это что? И какая функция вызывается, когда его пытаются _прочитать_? И что эта функция пытается _прочитать_ в первой строчке? И что вызывается, когда его пытаются _прочитать_? И что эта функция пытается _прочитать_ в первой строчке? И что вызывается, когда его пытаются _прочитать_? И что эта функция пытается _прочитать_ в первой строчке? ... Ну и так далее :slightly_smiling_face:

**Comment by uliss4ever** [comment]
ненене, только не в терновый куст, только не рекурсия)
Тогда у меня получается так. Вроде это не метод тыка (но всё-таки не точно). Ладно, завтра тогда)
![Снимок экрана 2021-03-09 в 20 24 51](https://user-images.githubusercontent.com/32711430/110512047-eb7bfd80-8115-11eb-990d-bc3108459c05.png)

Слушаю 3 лекцию: "если у вас есть большие сложности с пониманием HTML, задавайте вопросы, конечно, но особенно останавливаться я на этом не буду...". Где тут смайлик "застрелиться"?...

**Comment by Lex-2008** [comment]
Рекурсия переехала на 1 строчку ниже...

HTML можно посмотреть вот отсюда: https://www.w3schools.com/html/html_editors.asp
Ну а, вообще, HTML - это то, что видно в Хроме, когда видео через инструменты разработчика вытаскиваешь :)

**Comment by uliss4ever** [comment]
Мда... Но на все другие варианты пичарм ругается. Что-то видимо тут недопонимаю...

**Comment by uliss4ever** [comment]
> HTML можно посмотреть вот отсюда: https://www.w3schools.com/html/html_editors.asp
Ну а, вообще, HTML - это то, что видно в Хроме, когда видео через инструменты разработчика вытаскиваешь :)

Пасиб! То есть не страшно))

**Comment by uliss4ever** [comment]
С другой стороны... Про None мы уже сказали в сеттере. Так что пока наверно так. Но думаю дальше)

![Снимок экрана 2021-03-09 в 21 08 24](https://user-images.githubusercontent.com/32711430/110517450-ede15600-811b-11eb-95d5-81f56874ccd0.png)

**Comment by uliss4ever** [comment]
>self.driver (который без _) - это что? И какая функция вызывается, когда его пытаются прочитать? И что эта функция пытается прочитать в первой строчке?

Сеттер вызывается? И они с геттером зацикливаются, видимо

**Comment by Lex-2008** [comment]
> Так что пока наверно так.

Вот, правильно, молодец!

> Но думаю дальше)

Нет, хватит, остановись!! :rofl: всё уже правильно. Ну, кроме этого:

> И какая функция вызывается, когда его пытаются прочитать?
> 
> Сеттер вызывается?

Неет, Сеттер вызывается при попытке _установить_ ("set"), то есть _записать_. При попытке _прочитать_, то есть _взять_ ("get") значение, вызывается геттер.

**Comment by uliss4ever** [comment]
>Неет, Сеттер вызывается при попытке установить ("set"), то есть записать. При попытке прочитать, то есть взять ("get") значение, вызывается геттер.

Дада, точно. Но вот он пришел в геттер, а оттуда идёт в сеттер, чтобы узнать, что там установлено?

**Comment by Lex-2008** [comment]
> вот он пришел в геттер, а оттуда идёт в сеттер, чтобы узнать, что там установлено

Ненене, ты что! Компьютер не такой умный :) Чтобы _узнать_, что там установлено - ходят в геттер, чтобы _установить_ там что-то - ходят в сеттер!

**Comment by uliss4ever** [comment]
Смотри, геттер возвращает return self._driver
Но про self._driver, какое значение, у нас же только в сеттере сказано, в двух местах. То есть что вернуть геттеру, получается геттер всё равно узнаёт из сеттера? Или я опять всё перепутала?

**Comment by Lex-2008** [comment]
> То есть что вернуть геттеру, получается геттер всё равно узнаёт из сеттера?

В этом смысле да, но для этого сеттер должен был отработать заранее. Это как у Алисы: "Чтобы продать что-нибудь ненужное, надо сначала купить что-нибудь ненужное, а у нас денег нету." Почему это смешно? Потому что функция-метод "продавать"/отдавать (геттер) побежала в функцию-метод "покупать"/забирать (сеттер). Если бы Алиса сказала так, то всё было бы правильно: "Чтобы продать что-нибудь ненужное, надо **было** купить что-нибудь ненужное, _а мы это не сделали_". То есть, **сначала** покупатель/сеттер должен установить значение / положить в коробочку что-нибудь не нужное, и только **потом** продавец/геттер это значение оттуда берёт. В этом смысле да: что сеттер установил, то геттер и возвращает. Но если сеттер ничего не установил, то геттер к нему уже не побежит!

Ещё один пример: ребенок, которого научили командам "положи" и "дай". По команде "положи" он кладет то, что ему дали, себе в карман (сеттер), по команде "дай" - достает то, что у него в кармане и дает нам (геттер). Здесь так же: достать из кармана можно только то, что туда положили, но если ребенку дать команду "дай", когда там ничего не лежит, то он не должен внезапно выполнять команду "положи"!

**Comment by uliss4ever** [comment]
Мда... Но тогда проблема получается. Мы не можем ничего спросить у геттера, если до этого не дали это сеттеру. А всё, чего мы хотим, мы пишем в мэйне, так?
Или мы не всегда сеттером и геттером можем/должны пользоваться? И у нас ещё инит есть. И если в мэйне написано driver = PicleFileDriver("some.bin"), то у нас отрабатывает строчка в ините self.driver = driver и применяется этот пиклодрайвер. Да? А если мы в мэйне не сказали про драйвер ничего, то он у нас по умолчанию None и интерпретатор идёт куда? В геттер посмотреть или в сеттер установить? Плюс ещё в в сеттере есть не только про None, но и про value. В общем, что конкретно мне ещё не до конца ясно, я поняла.
